### ISSUE TITLE
kata: prepare transactions for dispatch / state contention resolution

### ISSUE THREAD

**Issue by hosie** [LF-Decentralized-Trust-labs/paladin#31]
Once a transaction has been initiated by a user and before it is ready to be dispatched ( submitted to the underlying base ledger), we need to determine 
 a) what sequencing caveats, if any, exist for each transaction.  This may be explicit pre-reqs defined by the user when the transaction was initiated or maybe subsequent to the assembly ( state selection) and endorsement loop handshake between other paladin entities
 b) which paladin node is responsible for submitting each sequence 

## Background
There are, at least, 3 stages required to process paladin transactions. Two of them are already relatively well understood:
  - Assembly stage is responsible for selecting which state(s) could be spent by this transaction and creating the new states produced by the transaction
  - Dispatch stage is responsible for assigning the transactions to a signing key's nonce sequence, signing the transaction and submitting it to the base ledger
  
In the simplest of cases, where there are no ordering requirements between the transactions, then the output of the assembly stage (a stream of assembled transactions) can be directly used as input to the dispatch stage. 
However, in reality, there are often ordering requirements and/or contention between transactions and we need an intermediary stage to resolve the ordering requirements and any contention between transactions. This task will implement that intermediary stage.

Ordering requirements can come from 2 sources
 - explicit ordering requirements declared by a user when submitting a transaction. This may be due to business logic for example, it may be desirable to execute a `sell` before a `buy` to ensure balance does not drop below a given threshold, or it may be desirable, for regulatory reasons, to stay below a given threshold and it is more important to execute the `buy` before the `sell`.
 - implicit ordering requirements determined by Paladin.  When selecting which state to spend for a given transaction, it may be necessary to select a state that is the output of another transaction that has not been confirmed yet.  Under normal conditions, this is perfectly valid provided that the minting transaction for that state does get ordered on the base ledger before the transaction that will spend that state.

Contention is the situation where 2, or more, transactions are assembled to spend the same state. Either transaction would be valid on its own but only one of them can be confirmed on the base ledger.  The other will be reverted and any transactions that depend on it will also be reverted.

The assembly stage runs as several independent single threaded processes. There is one assembly thread per domain per paladin node. This helps to minimise state contention between transactions( the case where 2 transactions are assembled to spend the same state). multiple transactions assembled by the same paladin will not be assembled to spend the same state because the assembler is single threaded, as soon as one transaction is assembled to spend a given state, the next iteration of the assembly thread will not see that state as available for selection.  However, 2 transactions assembled on 2 different paladin nodes are done so concurrently so there is a possibility that they will end up contending for the same state. 

The dispatch stage also runs as several independent single threaded processes. There is one dispatch thread per signing key. This assures ordering of transactions on the base ledger, at least, for all transactions processed by a given dispatcher.

To compliment `assembly` and `dispatch` stages, we need an `ordering` stage that will:
 - resolve ordering requirements between transactions
 - resolve contention between transactions by trigging re-assembly of transactions that are in contention
 - assign transactions to a dispatcher in a way that ensures that the ordering requirements are met

## Assumptions
In parallel to ordering, transactions must also be endorsed before they can move from assembly to dispatch.  Endorsement rules are domain specific so the proposal for ordering tries to be agnostic to the nuances of endorsement but does assume that the process of endorsement will ensure:
 1. spending rules are adhered to
 2. the masked data of all output states are valid for the given input states and transaction parameters
 3. at most one assembled transaction is endorsed to spend any input state

However, this proposal does *not* assume the process of endorsement will resolve any contention in the case where multiple transactions have been assembled to spend the same state.

In certain domains, it may be the case that there is a singleton endorser and/or exclusive spending rules which means that contention is naturally avoided.  However, there are domains where the spending rules allow any one of a number of parties to spend a given state and where endorsement requires multiple signatures.  So it is possible for the situation to arise where 2 transactions are assembled that are both perfectly valid as per assumption `1` & `2` above but neither of them achieve full endorsement as per assumption `3`. Further, because the individual signers for endorsement might learn about the transactions in different orders, some signers will endorse one transaction while other signers will endorse the other.  So neither transaction is the obvious rightful spender of that state.  

The following proposal assumes that the endorsement process does *not* attempt to resolve that situation and instead, it is the responsibility of the ordering stage to trigger transactions to be re-assembled. So this leads to a assumption that the endorsement stage will:
 4. in the case of transactions that could not achieve full endorsement due to assumption `3` , If some transactions are re-assembled to spend different states then endorsement will be re-evaluated for the remaining transaction attempting to spend a state.

## Proposal
 - The `sequencing` stage runs as several independent single threaded processes and each `sequencer` thread has an affinity with a `dispatcher` thread. There is one sequencer per domain per Paladin node.
 - Each sequencer maintains a graph of all transactions that are inflight. The graph is a directed acyclic graph where each node is a state or a transaction the edges are the relationships between states and transactions ( e.g. states are `spent by` or `minted by` transactions) ordering requirements between transactions.
 - Once transactions have been assembled by the assembly stage on a given Paladin node, it is added to the graph of the sequencer for that same domain and Paladin node.
 - The sequencer continuously evaluates whether the current graph can be divided into sub-graphs that are ready to be sorted (i.e. have zero contention between transactions and have full endorsement). If such a sub-graph can be isolated, then the sequencer will generate a topological sort of that sub-graph and hand it over to the dispatcher stage on the same Paladin node.
 - This handover from sequencer to dispatcher is a point of no return. Once a transaction has been assigned to a dispatcher, it will be submitted to the base ledger using the signing key and nonce sequence as chosen by that dispatcher.
 - Before that point of not return, there are many cases where it is possible for transactions to be removed from a graph i.e. the current sequencer will no longer be in control of sequencing for that transaction. e.g. Individual transactions may need to go back to the assembly stage (in case of state contention) or a whole graph, or sub graph may need to be merged into another graph being managed by another sequencer (in case of cross dependencies).
 - Graph analysis is complex event handler.  All sequencers on all Paladin nodes for a given domain will receive the same events (albeit in potentially different orders) and all arrive at the same conclusions with respect to eventually assigning each transaction to exactly one  `sequencer`.
 - The sequencer listens to events from sequencers for the same domain on other Paladin nodes and continuously analyses the graph to identify and resolve the following situations:
  - In the case where a transaction has an ordering dependency on a transaction that has already been endorsed and handed over to a dispatcher, then that transaction will be assigned to the sequencer that is coupled to that dispatcher.
  - In cases where a transaction has ordering dependencies on multiple inflight transactions that are being dispatched by multiple `dispatchers`, then that new transaction is held in the `blocked` stage until all bar one of those dependencies are confirmed on the base ledger. Once all bar one dependency transactions have been confirmed, then the rules per the previous bullet are enforced.
  - In cases where a transaction has an ordering requirement on multiple transactions that are currently under the purview of multiple different sequencers and none of those transactions have been endorsed and past the point of no return yet then, as above, the transaction will be marked as `blocked` until all bar one of the dependencies are confirmed. 
    - NOTE: there is a potential to improve the algorithm in this special case.  e.g. it may be possible to create a combined subgraph of all those dependencies and assign it to one sequencer.  However this could get really complex if you start to consider transitive dependencies and race conditions between the other sequencers getting past the point of no return while the combination of sub-graphs is being calculated.  Therefore, the proposal for this first iteration of the algorithm is to keep it simple and just block the transaction until all bar one of the dependencies are confirmed and potentially optimise it in future iterations.
  - In cases where multiple transactions are contending for the same state, 
    - In the case where that state is not yet confirmed, then as per the above rules, these transactions will eventually all end up under the purview of the sequencer responsible for creating that state. That sequencer will be responsible for resolving the contention. It will do so using a fair and deterministic function.
    - In the case where that state is confirmed, (assuming it is possible that the sequencer who was responsible for it may no longer be online) then then a fair and deterministic function will be used to assign all of those transactions to a single sequencer to resolve that contention. This is a function of the state hash and the output is a sequencer that is resident on a node where one of the contending transactions was assembled. NOTE: this may be a temporary assignment because before the point of no return, it may be the case that additional transactions are learned about that could change the outcome of the function to assign a sequencer to the contending transactions.
  - In cases where contention involves multiple states, 
    - if some of the states are not confirmed but the transactions creating them are they are all under the purview of the same sequencer, then all transactions in contention for those states will all be assigned to that sequencer.
    - if some of the states are not confirmed and the transactions creating them are currently under the purview of different sequencers, then all transactions in contention for those states will be marked as blocked as per the above rules.
    - if all of the states are confirmed, then a sequencer is chosen to resolve the contention using a function of the combined hashes of the states in contention. The output of this function is a sequencer that is resident on a node where one of the contending transactions was assembled. NOTE: this may be a temporary assignment because before the point of no return, it may be the case that additional transactions are learned about that could change the outcome of the function to assign a sequencer to the contending transactions.
      - NOTE: this may result in an undesirable situation where all transactions are re-assembled. Proposal is to accept that as a limitation in the particular edge case. Future iterations of this algorithm may improve upon this.

 -
